simulator/src/main.rs

```diff
@@ -0,0 +1,632 @@
+    events.insert(
+        Event::DataEvent(DataEvent {
+            event_type: DataEventType::GenerateBMS,
+            device: device.clone(),
+            timestamp: now + duration,
+            sequence: 1,
+        }),
+        duration,
+    );
+
+    let duration = DataEventType::GenerateIMU.duration();
+    events.insert(
+        Event::DataEvent(DataEvent {
+            event_type: DataEventType::GenerateIMU,
+            device: device.clone(),
+            timestamp: now + duration,
+            sequence: 1,
+        }),
+        duration,
+    );
+}
+
+pub async fn process_data_event(
+    event: &DataEvent,
+    events: &mut DelayQueue<Event>,
+    client: &mut SplitSink<Framed<TcpStream, LinesCodec>, String>,
+) -> Result<(), Error> {
+    let data = match event.event_type {
+        DataEventType::GenerateGPS => generate_gps_data(&event.device, event.sequence),
+        DataEventType::GenerateIMU => generate_imu_data(event.sequence),
+        DataEventType::GenerateVehicleData => generate_device_shadow_data(event.sequence),
+        DataEventType::GeneratePeripheralData => generate_peripheral_state_data(event.sequence),
+        DataEventType::GenerateMotor => generate_motor_data(event.sequence),
+        DataEventType::GenerateBMS => generate_bms_data(event.sequence),
+    };
+
+    let payload = serde_json::to_string(&data)?;
+    client.send(payload).await?;
+
+    let duration = event.event_type.duration();
+    let sequence = if event.sequence >= RESET_LIMIT { 1 } else { event.sequence + 1 };
+
+    events.insert(
+        Event::DataEvent(DataEvent {
+            sequence,
+            timestamp: event.timestamp + duration,
+            device: event.device.clone(),
+            event_type: event.event_type,
+        }),
+        duration,
+    );
+
+    Ok(())
+}
+
+async fn process_action_response_event(
+    event: &ActionResponseEvent,
+    client: &mut SplitSink<Framed<TcpStream, LinesCodec>, String>,
+) -> Result<(), Error> {
+    //info!("Sending action response {:?} {} {} {}", event.action_id, event.progress, event.status);
+    let payload = Payload {
+        stream: "action_status".to_string(),
+        device_id: None,
+        sequence: 0,
+        timestamp: SystemTime::UNIX_EPOCH.elapsed().unwrap().as_millis() as u64,
+        payload: json!({
+            "action_id": event.action_id,
+            "state": event.status,
+            "progress": event.progress,
+            "errors": []
+        }),
+    };
+
+    info!("Sending action response {} {} {}", event.action_id, event.progress, event.status);
+
+    let payload = serde_json::to_string(&payload)?;
+    client.send(payload).await?;
+    info!("Successfully sent action response");
+
+    Ok(())
+}
+
+pub async fn process_events(
+    events: &mut DelayQueue<Event>,
+    client: &mut SplitSink<Framed<TcpStream, LinesCodec>, String>,
+) -> Result<(), Error> {
+    if let Some(e) = events.next().await {
```

Comment:
What does `None` signify here? Early return or unwrap if this should never happen

Reply:
`None` in `DelayQueue` means empty schedule, which is possible if there are no events scheduled, highly unlikely, but possible, which is why I decided to keep the way that simulator used to handle it before, i.e. with an `if let`